{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geoengine as ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ConnectionError",
     "evalue": "HTTPConnectionPool(host='localhost', port=3030): Max retries exceeded with url: /api/anonymous (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4d166a0d30>: Failed to establish a new connection: [Errno 111] Connection refused'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionRefusedError\u001b[0m                    Traceback (most recent call last)",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connection.py:174\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 174\u001b[0m     conn \u001b[39m=\u001b[39m connection\u001b[39m.\u001b[39;49mcreate_connection(\n\u001b[1;32m    175\u001b[0m         (\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_dns_host, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mport), \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtimeout, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mextra_kw\n\u001b[1;32m    176\u001b[0m     )\n\u001b[1;32m    178\u001b[0m \u001b[39mexcept\u001b[39;00m SocketTimeout:\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/util/connection.py:95\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[39mif\u001b[39;00m err \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m---> 95\u001b[0m     \u001b[39mraise\u001b[39;00m err\n\u001b[1;32m     97\u001b[0m \u001b[39mraise\u001b[39;00m socket\u001b[39m.\u001b[39merror(\u001b[39m\"\u001b[39m\u001b[39mgetaddrinfo returns an empty list\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/util/connection.py:85\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     84\u001b[0m     sock\u001b[39m.\u001b[39mbind(source_address)\n\u001b[0;32m---> 85\u001b[0m sock\u001b[39m.\u001b[39;49mconnect(sa)\n\u001b[1;32m     86\u001b[0m \u001b[39mreturn\u001b[39;00m sock\n",
      "\u001b[0;31mConnectionRefusedError\u001b[0m: [Errno 111] Connection refused",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNewConnectionError\u001b[0m                        Traceback (most recent call last)",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    704\u001b[0m     conn,\n\u001b[1;32m    705\u001b[0m     method,\n\u001b[1;32m    706\u001b[0m     url,\n\u001b[1;32m    707\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    708\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    709\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    710\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    711\u001b[0m )\n\u001b[1;32m    713\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connectionpool.py:398\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    397\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 398\u001b[0m         conn\u001b[39m.\u001b[39;49mrequest(method, url, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mhttplib_request_kw)\n\u001b[1;32m    400\u001b[0m \u001b[39m# We are swallowing BrokenPipeError (errno.EPIPE) since the server is\u001b[39;00m\n\u001b[1;32m    401\u001b[0m \u001b[39m# legitimately able to close the connection after sending a valid response.\u001b[39;00m\n\u001b[1;32m    402\u001b[0m \u001b[39m# With this behaviour, the received response is still readable.\u001b[39;00m\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connection.py:239\u001b[0m, in \u001b[0;36mHTTPConnection.request\u001b[0;34m(self, method, url, body, headers)\u001b[0m\n\u001b[1;32m    238\u001b[0m     headers[\u001b[39m\"\u001b[39m\u001b[39mUser-Agent\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m _get_default_user_agent()\n\u001b[0;32m--> 239\u001b[0m \u001b[39msuper\u001b[39;49m(HTTPConnection, \u001b[39mself\u001b[39;49m)\u001b[39m.\u001b[39;49mrequest(method, url, body\u001b[39m=\u001b[39;49mbody, headers\u001b[39m=\u001b[39;49mheaders)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1282\u001b[0m, in \u001b[0;36mHTTPConnection.request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1281\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Send a complete request to the server.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1282\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_request(method, url, body, headers, encode_chunked)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1328\u001b[0m, in \u001b[0;36mHTTPConnection._send_request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1327\u001b[0m     body \u001b[39m=\u001b[39m _encode(body, \u001b[39m'\u001b[39m\u001b[39mbody\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m-> 1328\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mendheaders(body, encode_chunked\u001b[39m=\u001b[39;49mencode_chunked)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1277\u001b[0m, in \u001b[0;36mHTTPConnection.endheaders\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1276\u001b[0m     \u001b[39mraise\u001b[39;00m CannotSendHeader()\n\u001b[0;32m-> 1277\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_output(message_body, encode_chunked\u001b[39m=\u001b[39;49mencode_chunked)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1037\u001b[0m, in \u001b[0;36mHTTPConnection._send_output\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1036\u001b[0m \u001b[39mdel\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_buffer[:]\n\u001b[0;32m-> 1037\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(msg)\n\u001b[1;32m   1039\u001b[0m \u001b[39mif\u001b[39;00m message_body \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   1040\u001b[0m \n\u001b[1;32m   1041\u001b[0m     \u001b[39m# create a consistent interface to message_body\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:975\u001b[0m, in \u001b[0;36mHTTPConnection.send\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m    974\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mauto_open:\n\u001b[0;32m--> 975\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mconnect()\n\u001b[1;32m    976\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connection.py:205\u001b[0m, in \u001b[0;36mHTTPConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 205\u001b[0m     conn \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_new_conn()\n\u001b[1;32m    206\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_prepare_conn(conn)\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connection.py:186\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    185\u001b[0m \u001b[39mexcept\u001b[39;00m SocketError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m--> 186\u001b[0m     \u001b[39mraise\u001b[39;00m NewConnectionError(\n\u001b[1;32m    187\u001b[0m         \u001b[39mself\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mFailed to establish a new connection: \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m e\n\u001b[1;32m    188\u001b[0m     )\n\u001b[1;32m    190\u001b[0m \u001b[39mreturn\u001b[39;00m conn\n",
      "\u001b[0;31mNewConnectionError\u001b[0m: <urllib3.connection.HTTPConnection object at 0x7f4d166a0d30>: Failed to establish a new connection: [Errno 111] Connection refused",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/adapters.py:489\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 489\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    490\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    491\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    492\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    493\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    494\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    495\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    496\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    497\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    498\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    499\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    500\u001b[0m     )\n\u001b[1;32m    502\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/connectionpool.py:787\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    785\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> 787\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    788\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    789\u001b[0m )\n\u001b[1;32m    790\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[39mif\u001b[39;00m new_retry\u001b[39m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[39mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[39mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mIncremented Retry for (url=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m): \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPConnectionPool(host='localhost', port=3030): Max retries exceeded with url: /api/anonymous (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4d166a0d30>: Failed to establish a new connection: [Errno 111] Connection refused'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ge\u001b[39m.\u001b[39;49minitialize(\u001b[39m\"\u001b[39;49m\u001b[39mhttp://localhost:3030/api\u001b[39;49m\u001b[39m\"\u001b[39;49m, admin_token\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m8aca8875-425a-4ef1-8ee6-cdfc62dd7525\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/git/geoengine-python/geoengine/auth.py:199\u001b[0m, in \u001b[0;36minitialize\u001b[0;34m(server_url, credentials, token, admin_token)\u001b[0m\n\u001b[1;32m    186\u001b[0m \u001b[39m\u001b[39m\u001b[39m'''\u001b[39;00m\n\u001b[1;32m    187\u001b[0m \u001b[39mInitialize communication between this library and a Geo Engine instance\u001b[39;00m\n\u001b[1;32m    188\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    194\u001b[0m \u001b[39moptional .env file defining: GEOENGINE_EMAIL, GEOENGINE_PASSWORD, GEOENGINE_TOKEN\u001b[39;00m\n\u001b[1;32m    195\u001b[0m \u001b[39m'''\u001b[39;00m\n\u001b[1;32m    197\u001b[0m load_dotenv()\n\u001b[0;32m--> 199\u001b[0m Session\u001b[39m.\u001b[39msession \u001b[39m=\u001b[39m Session(server_url, credentials, token, admin_token)\n",
      "File \u001b[0;32m~/git/geoengine-python/geoengine/auth.py:88\u001b[0m, in \u001b[0;36mSession.__init__\u001b[0;34m(self, server_url, credentials, token, admin_token)\u001b[0m\n\u001b[1;32m     84\u001b[0m     session \u001b[39m=\u001b[39m req\u001b[39m.\u001b[39mget(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mserver_url\u001b[39m}\u001b[39;00m\u001b[39m/session\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m     85\u001b[0m                       headers\u001b[39m=\u001b[39m{\u001b[39m'\u001b[39m\u001b[39mAuthorization\u001b[39m\u001b[39m'\u001b[39m: \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mBearer \u001b[39m\u001b[39m{\u001b[39;00mos\u001b[39m.\u001b[39menviron\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mGEOENGINE_TOKEN\u001b[39m\u001b[39m\"\u001b[39m)\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m},\n\u001b[1;32m     86\u001b[0m                       timeout\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__timeout)\u001b[39m.\u001b[39mjson()\n\u001b[1;32m     87\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m---> 88\u001b[0m     session \u001b[39m=\u001b[39m req\u001b[39m.\u001b[39;49mpost(\u001b[39mf\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m{\u001b[39;49;00mserver_url\u001b[39m}\u001b[39;49;00m\u001b[39m/anonymous\u001b[39;49m\u001b[39m'\u001b[39;49m, timeout\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__timeout)\u001b[39m.\u001b[39mjson()\n\u001b[1;32m     90\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39merror\u001b[39m\u001b[39m'\u001b[39m \u001b[39min\u001b[39;00m session:\n\u001b[1;32m     91\u001b[0m     \u001b[39mraise\u001b[39;00m GeoEngineException(session)\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/api.py:115\u001b[0m, in \u001b[0;36mpost\u001b[0;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[1;32m    103\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(url, data\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, json\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    104\u001b[0m \u001b[39m    \u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a POST request.\u001b[39;00m\n\u001b[1;32m    105\u001b[0m \n\u001b[1;32m    106\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    113\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 115\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m\"\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m\"\u001b[39;49m, url, data\u001b[39m=\u001b[39;49mdata, json\u001b[39m=\u001b[39;49mjson, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    582\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    583\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[1;32m    584\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    585\u001b[0m }\n\u001b[1;32m    586\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 587\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    589\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/sessions.py:701\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    698\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    700\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    703\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    704\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m~/git/geoengine-python/env/lib/python3.10/site-packages/requests/adapters.py:565\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    561\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(e\u001b[39m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    562\u001b[0m         \u001b[39m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[1;32m    563\u001b[0m         \u001b[39mraise\u001b[39;00m SSLError(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[0;32m--> 565\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    567\u001b[0m \u001b[39mexcept\u001b[39;00m ClosedPoolError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    568\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(e, request\u001b[39m=\u001b[39mrequest)\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPConnectionPool(host='localhost', port=3030): Max retries exceeded with url: /api/anonymous (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4d166a0d30>: Failed to establish a new connection: [Errno 111] Connection refused'))"
     ]
    }
   ],
   "source": [
    "ge.initialize(\"http://localhost:3030/api\", admin_token=\"8aca8875-425a-4ef1-8ee6-cdfc62dd7525\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "volumes = ge.volumes()\n",
    "volumes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "start_time = \"1984-01-01T00:00:00Z\"\n",
    "end_time = \"2023-01-02T00:00:00Z\"\n",
    "\n",
    "data_path_in_volume = \"respect/ecmwf_era5\"\n",
    "\n",
    "pressure_levels = [\n",
    "            '300', '350', '400',\n",
    "            '450', '500', '550',\n",
    "            '600', '650', '700',\n",
    "            '750', '775', '800',\n",
    "            '825', '850', '875',\n",
    "            '900', '925', '950',\n",
    "            '975', '1000',\n",
    "        ]\n",
    "\n",
    "dataset_ids_air_temperature = {\n",
    "    '300': \"7c9cc95e-b3ed-461a-ae0e-21b2520d3391\",\n",
    "    '350': \"3f76b90b-2eff-48fa-bb63-c6d5078392e7\",\n",
    "    '400': \"105ad9f8-67f3-4cb0-b439-a92fb8443d22\",\n",
    "    '450': \"37042fee-e7ff-40dd-9195-a04708e31a97\",\n",
    "    '500': \"35e37da3-8c09-4471-988b-8a78cd4515b0\",\n",
    "    '550': \"28a33e1a-9c7c-4d74-8b48-3941f4ac94e6\",\n",
    "    '600': \"3fd40dd3-c89d-4ccc-a162-6ada9383735a\",\n",
    "    '650': \"0a0acfdb-ac58-4d94-a07c-76e8db0c8c2f\",\n",
    "    '700': \"88673023-2087-4282-acc8-8214bf2fb557\",\n",
    "    '750': \"58fd9e61-1a4f-4b45-badd-c6a8f013de8a\",\n",
    "    '775': \"68df526b-1dfc-49c4-aef7-035755ac491f\",\n",
    "    '800': \"745398ab-c114-442d-af00-e6faeeb43f01\",\n",
    "    '825': \"c6dd1a0f-7d8e-4435-bd59-4f6ea130f76e\",\n",
    "    '850': \"8ce7ea1a-ec10-4048-b248-2570f304afe7\",\n",
    "    '875': \"5ada9cec-1983-42a7-99d7-a41f2cb9963c\",\n",
    "    '900': \"fd72b75f-7c67-4dbd-b82c-c313bfef2e18\",\n",
    "    '925': \"ce768c83-b7a4-4b72-8a24-4a793eb53ccb\",\n",
    "    '950': \"8835b0f1-6106-4ee2-b91b-5a136142fc58\",\n",
    "    '975': \"d0135d9c-d1ce-495d-9d6e-41462ce93150\",\n",
    "    '1000': \"06589b99-c9e4-4bb3-bde2-e425745c6255\",\n",
    "}\n",
    "\n",
    "dataset_ids_u_component_of_wind = {\n",
    "    '300': \"17310fd0-99af-4f68-b252-e04f3bf306b0\",\n",
    "    '350': \"fe5f1c0f-b8c3-4b71-b9d3-a7a22b1f2462\",\n",
    "    '400': \"5069649a-c151-4ab5-b09a-b9195577fff5\",\n",
    "    '450': \"4d1b5b5c-fb51-4fb2-9620-5397b7eef98a\",\n",
    "    '500': \"154e549f-677f-4478-aa6d-dd68e1dc9984\",\n",
    "    '550': \"95f4b43d-b84a-4eae-bb31-7fc4a64354e2\",\n",
    "    '600': \"cd70021c-8bf8-41d2-9c2f-cc9ca616a351\",\n",
    "    '650': \"6835de67-bd6a-4f86-8e69-29f5eff464b6\",\n",
    "    '700': \"c00265c5-0ac6-4f23-b2bb-599a66e4d72a\",\n",
    "    '750': \"7ee53aa9-c3e0-404a-b181-c5cbbe5cb268\",\n",
    "    '775': \"26636f28-d948-4ae3-8252-78ab732bb13b\",\n",
    "    '800': \"2c42f02b-64e5-477e-b2d5-082d8e0ce11a\",\n",
    "    '825': \"157ca7a2-0e87-4e8e-a748-59b8b79aa847\",\n",
    "    '850': \"6e62e460-a662-4843-ae6b-4edb065b627a\",\n",
    "    '875': \"ea6aafff-72ed-4c49-8c98-be6dc0410be1\",\n",
    "    '900': \"9a91fa27-36c0-4256-855c-d97ba03eb0ec\",\n",
    "    '925': \"7610c19f-647c-4309-aa0e-9ae429aa8e12\",\n",
    "    '950': \"d33d93f9-5c57-41e0-a1e2-8d504f0f4a5e\",\n",
    "    '975': \"d907a0b6-79a5-4d4d-967b-dedf3d1d9835\",\n",
    "    '1000': \"5f7686b8-3209-4fe4-824d-2c5d12ecd591\",\n",
    "}\n",
    "\n",
    "dataset_ids_v_component_of_wind = {\n",
    "    '300': \"3aeaa3a2-2e35-490d-b1ae-645c4f9be8c6\",\n",
    "    '350': \"4016793a-e719-4913-ae74-46ba88cb7b84\",\n",
    "    '400': \"18135482-ec47-4b06-8209-f47ae7e67309\",\n",
    "    '450': \"216c63ed-4aff-4361-beb5-880e211bdfb0\",\n",
    "    '500': \"55cb95f9-b79f-44c1-9b74-5d95b9bb9893\",\n",
    "    '550': \"39fc1116-57a3-48e8-bd41-5e8f55dd5baa\",\n",
    "    '600': \"e779580b-34f1-419f-a08f-d8360bb05386\",\n",
    "    '650':\"ad79ffa7-0e8c-49a3-97d0-3db61e7fedda\",\n",
    "    '700':  \"f18ead07-3740-46d9-86dc-5a2e11b2f2df\",\n",
    "    '750': \"1bbd0b28-6ec3-4dab-b7c1-45ad2fbdf4ca\",\n",
    "    '775':\"4fbda734-a8f7-40cb-999f-cd762f3e350d\",\n",
    "    '800': \"a7290e1e-a090-4daf-8059-467aed5f235d\",\n",
    "    '825':  \"dc7b52ed-53a8-42f3-8791-2954203cd179\",\n",
    "    '850': \"2fbb61b6-f95b-4d6f-9ba2-92ab551090d1\",\n",
    "    '875':  \"5fcbbe4c-a702-44cf-91a3-f23065f928e5\",\n",
    "    '900':\"040edc42-c051-4c39-9982-7ca802c20614\",\n",
    "    '925':\"6373c8e8-e17b-4803-9be0-2e0c2d7404c0\",\n",
    "    '950': \"bd6566e4-0018-4423-92dd-3763378de9e4\",\n",
    "    '975': \"14bffda1-551b-4444-8c29-cc6f7073dbfa\",\n",
    "    '1000': \"bfca3186-0c81-44c7-9b2f-727c6acdd122\",\n",
    "}\n",
    "\n",
    "dataset_ids_specific_humidity = {\n",
    "    '300': \"79ff1546-07fe-4cb6-950a-13597efd6762\",\n",
    "    '350': \"95a2c739-fa5c-4b55-a404-f6eda5f5f335\",\n",
    "    '400': \"e9ce8747-684d-4ee5-8aed-c775402889c4\",\n",
    "    '450': \"2285fc68-95af-4606-b05d-b17773993727\",\n",
    "    '500': \"f3189fe1-1a28-4985-8eb3-bbac3887e1f5\",\n",
    "    '550': \"ff82f4c2-1749-4c8e-8556-583514622b19\",\n",
    "    '600': \"48945bba-88f5-40df-ab41-76c0969f9c9b\",\n",
    "    '650': \"3399ff2f-d95a-444f-a175-3723eeac910b\",\n",
    "    '700': \"3859e81e-dda8-4b8e-ab36-825f7c9ca734\",\n",
    "    '750': \"da879658-beb9-4f42-aee6-6df31e5bbb72\",\n",
    "    '775': \"5573e5ba-9fb8-41e7-8db1-2c31a9b5736d\",\n",
    "    '800': \"9817b367-acf9-44c4-98cc-b74a09c075e7\",\n",
    "    '825': \"f07d897a-f196-467c-a9e3-891a827276bb\",\n",
    "    '850': \"dfc8cd1b-c30b-473d-aa0b-062d9b0f5944\",\n",
    "    '875': \"dd88a94a-1911-499e-8d6d-ee0dd63dffa1\",\n",
    "    '900': \"39c8aed1-e14a-4b04-a1af-20bd2de93d88\",\n",
    "    '925': \"aaa3ca70-f9e6-45fc-a939-a6071590d010\",\n",
    "    '950': \"0e76a9c7-0000-4b9a-8ef2-5f146eaea368\",\n",
    "    '975': \"3f18fafc-658b-4a4b-9a94-51ba6b6e9667\",\n",
    "    '1000': \"94f482a7-ebc8-4466-93d6-f21e552f6c57\",\n",
    "}\n",
    "\n",
    "dataset_ids_relative_humidity = {\n",
    "    '300': \"63f2df9b-43c3-4255-bf47-24bff24bd0fc\",\n",
    "    '350':  \"82845451-4fb8-46bf-be9e-e600891845d9\",\n",
    "    '400': \"cc4027af-9908-4f22-b603-c2b5d3928099\",\n",
    "    '450':  \"84caac6f-965c-45e4-b45f-d38d5eaa970f\",\n",
    "    '500': \"06641fdc-4023-4b19-be57-72100955f362\",\n",
    "    '550': \"43b63466-01e3-4da0-b6f9-7c306d169949\",\n",
    "    '600':  \"fbf54507-466c-4226-82c3-34aabf4a3440\",\n",
    "    '650':  \"d82b6f33-72ce-4de7-8eaf-210fa7161dbe\",\n",
    "    '700': \"8bf9849d-f106-463c-9fd4-db07487e523f\",\n",
    "    '750':\"cc1e6d57-478b-4950-8d6f-3ef61c82ebde\",\n",
    "    '775': \"a7c75061-8dc9-4a6a-aa1a-78dd7c1fd429\",\n",
    "    '800':\"329381d0-0345-4b9d-af3b-b8e53b571a1d\",\n",
    "    '825': \"98ddba41-e7d1-4a33-a132-aa541f912b67\",\n",
    "    '850':\"f26cb97e-0180-4c57-83bf-349dd36147f0\",\n",
    "    '875':  \"ed61b695-3caf-4dce-b8ea-8bec2fdde258\",\n",
    "    '900':  \"bb82a48f-c959-4752-959a-89be9b72c31e\",\n",
    "    '925': \"a4d24f45-1ed8-40f0-9a96-c8b15c0051e5\",\n",
    "    '950':  \"aae1a654-30c2-4669-8821-e107e7e879a1\",\n",
    "    '975':  \"ea20e7e7-1fd6-4085-9856-5541686a9db9\",\n",
    "    '1000':\"9a0793c3-e63a-4cae-8cba-0933087a0ba7\",\n",
    "}\n",
    "\n",
    "variables_with_props = {\n",
    "    'air_temperature': {\n",
    "        \"file_suffix\": \"air_temperature\",\n",
    "        \"name\": \"Air Temperature\",\n",
    "        \"unit\": \"K\",\n",
    "        \"measurement\": \"temperature\",\n",
    "        \"level_ids\": dataset_ids_air_temperature,\n",
    "    },\n",
    "    'eastward_wind': {\n",
    "        \"file_suffix\": \"eastward_wind\",\n",
    "        \"name\": \"Eastward Wind\",\n",
    "        \"unit\": \"m/s\",\n",
    "        \"measurement\": \"u_component_of_wind\",\n",
    "        \"level_ids\": dataset_ids_u_component_of_wind,\n",
    "    },\n",
    "    'northward_wind': {\n",
    "        \"file_suffix\": \"northward_wind\",\n",
    "        \"name\": \"Northward Wind\",\n",
    "        \"unit\": \"m/s\",\n",
    "        \"measurement\": \"v_component_of_wind\",\n",
    "        \"level_ids\": dataset_ids_v_component_of_wind,\n",
    "    },\n",
    "    'specific_humidity': {\n",
    "        \"file_suffix\": \"specific_humidity\",\n",
    "        \"name\": \"Specific Humidity\",\n",
    "        \"unit\": \"kg/kg\",\n",
    "        \"measurement\": \"specific_humidity\",\n",
    "        \"level_ids\": dataset_ids_specific_humidity,\n",
    "    },\n",
    "    'relative_humidity': {\n",
    "        \"file_suffix\": \"relative_humidity\",\n",
    "        \"name\": \"Relative Humidity\",\n",
    "        \"unit\": \"%\",\n",
    "        \"measurement\": \"relative_humidity\",\n",
    "        \"level_ids\": dataset_ids_relative_humidity,\n",
    "    },\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_props_and_meta(data_path_in_volume, name, file_name_suffix, level, dataset_id, start_time, end_time, measurement, unit):\n",
    "\n",
    "  upper_left_x_coord = -82.125\n",
    "  upper_left_y_coord = 3.125\n",
    "  pixel_size = 0.25\n",
    "  with_in_pixels = 33\n",
    "  height_in_pixels = 33\n",
    "\n",
    "  lower_right_x_coord = upper_left_x_coord + (pixel_size * with_in_pixels)\n",
    "  lower_right_y_coord = upper_left_y_coord - (pixel_size * height_in_pixels)\n",
    "  \n",
    "\n",
    "  unscaled_colorizer = ge.LinearGradientColorizer(\n",
    "          breakpoints = [\n",
    "            ge.ColorBreakpoint(-32766, (68, 1, 84, 255)),\n",
    "            ge.ColorBreakpoint( -32307.5, (33, 145, 140, 255)),\n",
    "            ge.ColorBreakpoint(-31849, (253, 231, 37, 255))\n",
    "          ],\n",
    "          no_data_color =(0, 0, 0, 0),\n",
    "          default_color = (0, 0, 0, 0)\n",
    "        )\n",
    "\n",
    "  unscaled_symbology = ge.RasterSymbology(\n",
    "        colorizer = unscaled_colorizer\n",
    "      )\n",
    "\n",
    "  result_descriptor = ge.RasterResultDescriptor(\n",
    "    data_type = \"I16\",\n",
    "    spatial_reference = \"EPSG:4326\",    \n",
    "    time_bounds=ge.TimeInterval(start_time, end_time),\n",
    "    spatial_resolution=ge.SpatialResolution(0.25, 0.25),\n",
    "    spatial_bounds=ge.BoundingBox2D(upper_left_x_coord, lower_right_y_coord, lower_right_x_coord, upper_left_y_coord),\n",
    "    measurement = ge.ContinuousMeasurement(unit = unit, measurement = measurement)\n",
    "  )\n",
    "\n",
    "  era5_add_data_properties = ge.DatasetProperties(\n",
    "      dataset_id = dataset_id,\n",
    "      name = f'ECMWF ERA5 {name} Level {level}',\n",
    "      description = f'ECMWF ERA5 {name} Level {level}',\n",
    "      source_operator = \"GdalSource\",\n",
    "      provenance = ge.Provenance(\n",
    "        citation = \"Hersbach, H., Bell, B., Berrisford, P., Biavati, G., Hor\\u00e1nyi, A., Mu\\u00f1oz Sabater, J., Nicolas, J., Peubey, C., Radu, R., Rozum, I., Schepers, D., Simmons, A., Soci, C., Dee, D., Th\\u00e9paut, J-N. (2018): ERA5 hourly data on pressure levels from 1959 to present. Copernicus Climate Change Service (C3S) Climate Data Store (CDS). (Accessed on < DD-MMM-YYYY >), 10.24381/cds.bd0915c6\",\n",
    "        license = \"https://cds.climate.copernicus.eu/api/v2/terms/static/licence-to-use-copernicus-products.pdf\",\n",
    "        uri = \"https://doi.org/10.24381/cds.bd0915c6\"\n",
    "      ),\n",
    "      symbology = unscaled_symbology,\n",
    "  )\n",
    "\n",
    "  era5_add_data_meta_data = ge.api.GdalMetaDataRegular({\n",
    "      \"type\": \"GdalMetaDataRegular\",\n",
    "      \"resultDescriptor\": result_descriptor.to_api_dict(),\n",
    "      \"params\": ge.api.GdalDatasetParameters({\n",
    "        \"filePath\": f'{data_path_in_volume}/%_YEAR_%/%_MONTH_%/%_START_DATE_%_{file_name_suffix}_{level}.tif',\n",
    "        \"rasterbandChannel\": 1,\n",
    "        \"geoTransform\": {\n",
    "          \"originCoordinate\": {\n",
    "            \"x\": upper_left_x_coord,\n",
    "            \"y\": upper_left_y_coord\n",
    "          },\n",
    "          \"xPixelSize\": pixel_size,\n",
    "          \"yPixelSize\": -1.0 * pixel_size\n",
    "        },\n",
    "        \"width\": with_in_pixels,\n",
    "        \"height\": height_in_pixels,\n",
    "        \"fileNotFoundHandling\": \"NoData\",\n",
    "        \"noDataValue\": -32767\n",
    "      }),\n",
    "      \"timePlaceholders\": {\n",
    "        \"%_START_DATE_%\": {\n",
    "          \"format\": {\n",
    "            'fmt': \"%Y%m%d_%H%M\",\n",
    "            \"has_tz\": False,\n",
    "            \"has_time\": True,\n",
    "            },\n",
    "          \"reference\": \"start\"\n",
    "        },\n",
    "        \"%_YEAR_%\": {\n",
    "          \"format\": {\n",
    "            'fmt':  \"%Y\",\n",
    "            \"has_tz\": False,\n",
    "            \"has_time\": True,\n",
    "          },\n",
    "          \"reference\": \"start\"\n",
    "        },\n",
    "        \"%_MONTH_%\": {\n",
    "          \"format\": {\n",
    "            'fmt': \"%m\",\n",
    "            \"has_tz\": False,\n",
    "            \"has_time\": True,\n",
    "          },\n",
    "          \"reference\": \"start\"\n",
    "        }\n",
    "      },\n",
    "      \"dataTime\": {\n",
    "        \"start\": start_time,\n",
    "        \"end\": end_time\n",
    "      },\n",
    "      \"step\": {\n",
    "        \"granularity\": \"hours\",\n",
    "        \"step\": 1\n",
    "      }\n",
    "    })\n",
    "\n",
    "\n",
    "  return (era5_add_data_properties, era5_add_data_meta_data) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "for variable_key in variables_with_props:\n",
    "\n",
    "    variable = variables_with_props[variable_key]\n",
    "    \n",
    "    file_suffix = variable['file_suffix']\n",
    "    v_name = variable['name']\n",
    "\n",
    "    out_dir = f'./test_data/era5/{file_suffix}/'\n",
    "\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)\n",
    "\n",
    "    for level in pressure_levels:\n",
    "        dataset_id = variable['level_ids'][level]\n",
    "        file_suffix = variable['file_suffix']\n",
    "        v_name = variable['name']\n",
    "        v_unit = variable['unit']\n",
    "        v_measurement = variable['measurement']\n",
    "\n",
    "        (a,m) = generate_props_and_meta(data_path_in_volume, v_name, file_suffix, level, dataset_id, start_time, end_time, v_measurement, v_unit)\n",
    "\n",
    "        definition = ge.api.DatasetDefinition({\n",
    "            'properties': a.to_api_dict(),\n",
    "            'metaData': m\n",
    "        })\n",
    "\n",
    "        with open(os.path.join(out_dir, f\"{file_suffix}_level_{level}.json\"), \"w\") as out_file:\n",
    "            json.dump(definition, out_file, indent=2)\n",
    "            out_file.write(\"\\n\")\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "volume = volumes[0]\n",
    "\n",
    "variable_name = 'air_temperature'\n",
    "variable = variables_with_props[variable_name]\n",
    "\n",
    "for level in pressure_levels:\n",
    "  dataset_id = dataset_ids_air_temperature[level]\n",
    "  (a,m) = generate_props_and_meta(volume, data_path_in_volume, variable['name'], variable['file_suffix'], level, dataset_id, start_time, end_time)\n",
    "  print(a.to_api_dict())\n",
    "  print(m)\n",
    "  dataset_id_returned = ge.add_public_raster_dataset(volumes[0].name, a, m)\n",
    "  print(dataset_id, dataset_id_returned)\n",
    "  assert str(dataset_id) == str(dataset_id_returned) # make sure the dataset id returned is the same as the one we wanted\n",
    "\n",
    "  \n",
    "\n",
    "  layer_id = layer_ids_air_temperature[level]\n",
    "  layer = ge.Layer(\n",
    "    layer_id=layer_id,\n",
    "    workflow=make_raster_workflow(dataset_id),\n",
    "    name=f\"ECMWF ERA5 {variable['name']} Level {level}\",\n",
    "    description=f\"ECMWF ERA5 {variable['name']} Level {level}\",\n",
    "    symbology=None,\n",
    "    meta_data={},\n",
    "    properties=[],\n",
    "  )\n",
    "  print(layer.to_api_dict())\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layer_ids_air_temperature = {\n",
    "    '300': \"dd3be565-c590-4155-9aea-662f9adfaa52\",\n",
    "    '350': \"5ef4d541-ae7f-4ef9-85e5-8940cbed2faf\",\n",
    "    '400': \"023c21a5-8d71-4582-83d6-75558361e043\",\n",
    "    '450': \"b99bb974-56d8-42dc-ae56-ead202f7a469\",\n",
    "    '500': \"c8c22bfb-d743-40ec-8c06-081b8cd3756c\",\n",
    "    '550': \"df6a2f62-aa85-4167-8c91-4a353b21eb19\",\n",
    "    '600': \"784fd7bf-479c-4d53-9210-21342f0d8079\",\n",
    "    '650': \"bc34430a-4573-4be8-bdef-011e3215500d\",\n",
    "    '700': \"2e902e7a-8bd6-4e39-9af0-294c9105ec15\",\n",
    "    '750': \"e5699818-93d7-424e-b042-a4a325a46a97\",\n",
    "    '775': \"cf214bfd-c40c-4491-87d3-1b71f8252199\",\n",
    "    '800': \"32d1ba9e-1a72-4030-9b1b-2ca71f55ecaf\",\n",
    "    '825': \"10ece546-882d-4f4e-8980-d9cea38269ff\",\n",
    "    '850': \"8607429c-61fe-4db9-80d2-40b1177836c5\",\n",
    "    '875': \"4dfd412a-80a4-439a-ab10-f85ba0ebae2e\",\n",
    "    '900': \"ae861b63-66b3-43ce-8a60-f05884b4b770\",\n",
    "    '925': \"6c649514-2b81-49af-8e48-1b6d81ac171f\",\n",
    "    '950': \"89d5a6cc-8bb7-496d-b06f-19c97147a11e\",\n",
    "    '975': \"efbbf7e6-b7d0-45fe-8d12-4077abbbc6eb\",\n",
    "    '1000': \"0434f5d7-7c17-41b8-a789-49a0449a744e\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for level in pressure_levels:\n",
    "  dataset_id = dataset_ids_air_temperature[level]\n",
    "  ge.delete_dataset(dataset_id)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a10ca932ff03f4c637e11f8207fe4e003ad6532d0070b911d37acf855f595fc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
